### PR TITLE
Enable multihead atten

### PR DESCRIPTION
**Pull request by hubertlu-tw**
### Installation:
`python setup.py install --cpp_ext --cuda_ext --distributed_adam --xentropy --deprecated_fused_adam --fast_multihead_attn 2>&1 |  tee ../apex_build_mha.log
`

### Unit tests
`cd tests/L0/ && bash run_rocm.sh 2>&1 | tee ../../apex_unittests.txt`
`cd tests/distributed/ && bash run_rocm_distributed.sh 2>&1 | tee ../../apex_distributed_unittests.txt`

Notice that I have built the apex on a MI200 server and confirmed that there is no new failing unit tests introduced. 

### Unit tests for the extension
(https://github.com/ROCmSoftwarePlatform/apex/tree/dev/hubertlu/multihead_attn/apex/contrib/test/multihead_attn)
<html xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=PowerPoint.Slide>
<meta name=Generator content="Microsoft PowerPoint 15">

</head>

<body>

  | CUDA | ROCm
-- | -- | --
test_encdec_multihead_attn.py | PASS | PASS
test_encdec_multihead_attn_norm_add.py | PASS | PASS
test_fast_self_multihead_attn_bias.py | PASS | **FAILED**
test_mha_fused_softmax.py | PASS | PASS
test_self_multihead_attn.py | PASS | PASS
test_self_multihead_attn_norm_add.py | FAILED | FAILED

</body>

</html>

Notice that the features for the failed test in test_fast_self_multihead_attn_bias.py on ROCm are not used by MLPerf team. We will need to root cause it later. In addition, the failed test in test_self_multihead_attn_norm_add.py is due to a missing 1 required positional argument in the upstream (NVIDIA) script.

Lastly, the current CI checks do not run the unit tests for extensions (such as groupbn, layer_norm, multihead_attn, and test_label_smoothing.py). We will need to add them to our CI checks later.

### CODE REVIEW

**Review by athitten — APPROVED** [review]
Everything looks good to me.

File: setup.py [hunk 1]
```diff
@@ -34,6 +34,11 @@ def check_if_rocm_pytorch():
 
 IS_ROCM_PYTORCH = check_if_rocm_pytorch()
 
+if IS_ROCM_PYTORCH:
+    rocm_include_dirs = ["/opt/rocm/include/hiprand", "/opt/rocm/include/rocrand"]
+else:
+    rocm_include_dirs = []
+
```
Comment (jeffdaily) [inline setup.py:41]:
```suggestion
```

The variable `rocm_include_dirs` does not appear to be used.  Can remove.

File: apex/contrib/csrc/multihead_attn/encdec_multihead_attn_cuda.cu [hunk 2]
```diff
@@ -1,12 +1,15 @@
 #include <vector>
 #include <iostream>
-
+//below lines enable hip float to half conversion which are disabled by default in hip_fp16.h
+#undef __HIP_NO_HALF_OPERATORS__
+#undef __HIP_NO_HALF_CONVERSIONS__
+//#endif
```
Comment (jeffdaily) [inline apex/contrib/csrc/multihead_attn/encdec_multihead_attn_cuda.cu:6]: Not sure if these are needed.  Might be redundant.  The setup.py adds `-U` for these symbols.

Comment (jeffdaily) [reply apex/contrib/csrc/multihead_attn/encdec_multihead_attn_cuda.cu:6]: There are a few other source files where these undefs are added.  Please check whether any/all of them are needed, or if the setup.py is sufficient.

Comment (hubertlu-tw) [reply apex/contrib/csrc/multihead_attn/encdec_multihead_attn_cuda.cu:6]: Thanks. I will look into this.

Comment (jithunnair-amd) [reply apex/contrib/csrc/multihead_attn/encdec_multihead_attn_cuda.cu:6]: @hubertlu-tw I think we never addressed this redundancy? Jeff had a point that the setup.py already does the required undefines: https://github.com/ROCmSoftwarePlatform/apex/blob/62f06964d8716deb681b39ac5cc24c2cbf2d53f2/setup.py#L400

**Pull request merged**